lib/mongo/server/connectable.rb

```diff
@@ -88,10 +88,11 @@ def ensure_connected
         ensure_same_process!
         connect!
         begin
-          yield socket
-        rescue Exception => e
-          disconnect!
-          raise e
+          res = yield socket
```

Comment:
We have to be sure that the connection is not checked in before the socket is closed. (i.e. that disconnect! is called here before checkin(connection) is called in connection_pool. It makes sense that the ensure block here would be called before the ensure block in with_connection of connection_pool, but I'll do some tests now to be 100% sure.